File: 07-security/08-crypto-and-rate-limiting/golang/crypto_ratelimit.go
// Layer 7 · Topic 8 — Crypto hygiene and rate limiting (Go).
//
// `GOFLAGS=-mod=mod GOPROXY=off go run crypto_ratelimit.go` (x/crypto is
// cached). Go's stdlib acknowledges constant-time as a CATEGORY: crypto/subtle
// has ConstantTimeCompare, ConstantTimeSelect, ConstantTimeByteEq. Hashing is
// golang.org/x/crypto/argon2; rate limiting would be golang.org/x/time/rate
// (in-process, so the distributed-state problem is yours -- which Part C shows).
//
// Three parts, measured at runtime: (A) sha256 vs argon2id verify/sec + the
// crack-time model; (B) the timing signal, naive short-circuit vs
// subtle.ConstantTimeCompare; (C) rate-limiting effective limit and fake fixes.
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"fmt"
	"time"

	"golang.org/x/crypto/argon2"
)

func partA() {
	fmt.Println("A. Hash cost (verifications/sec, measured)")
	pw := []byte("correct horse battery staple")

	reps := 500000
	t0 := time.Now()
	for i := 0; i < reps; i++ {
		sha256.Sum256(pw)
	}
	shaVps := float64(reps) / time.Since(t0).Seconds()
	fmt.Printf("   sha256          %14.0f verify/sec\n", shaVps)

	// argon2id at the OWASP baseline: m=19456 KiB, t=2, p=1.
	salt := make([]byte, 16)
	rand.Read(salt)
	reps = 20
	t0 = time.Now()
	for i := 0; i < reps; i++ {
		argon2.IDKey(pw, salt, 2, 19456, 1, 32)
	}
	argVps := float64(reps) / time.Since(t0).Seconds()
	fmt.Printf("   argon2id(19MiB) %14.1f verify/sec\n", argVps)

	N, K := 10000.0, 1000000.0
	fmt.Printf("   crack-time model: attacker rig N=%.0fx, list K=%.0f candidates\n", N, K)
	fmt.Printf("      sha256:   %.6f s to first crack\n", K/(shaVps*N))
	fmt.Printf("      argon2id: %.1f s to first crack  -- ~%.0fx slower per verify than sha256\n",
		K/(argVps*N), shaVps/argVps)
	fmt.Println()
}

var sink int // prevents dead-code elimination of timed compares

func naiveEq(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] { // short-circuit -> timing depends on the secret
			return false
		}
	}
	return true
}

func partB() {
	fmt.Println("B. Timing signal: naive short-circuit vs constant-time")
	secret := make([]byte, 32)
	rand.Read(secret)

	candidate := func(matching int) []byte {
		c := make([]byte, 32)
		rand.Read(c)
		copy(c[:matching], secret[:matching])
		if matching < 32 {
			c[matching] = secret[matching] ^ 0xFF
		}
		return c
	}
	// A single 32-byte compare is faster than the ns timer, so average ns/op
	// over a large batch; `sink` stops the compiler eliminating the call.
	avgNs := func(fn func(a, b []byte) int, cand []byte, reps int) float64 {
		t0 := time.Now()
		for i := 0; i < reps; i++ {
			sink += fn(secret, cand)
		}
		return float64(time.Since(t0).Nanoseconds()) / float64(reps)
	}
	naiveWrap := func(a, b []byte) int {
		if naiveEq(a, b) {
			return 1
		}
		return 0
	}

	fmt.Println("   matching leading bytes ->        avg ns/op")
	for _, v := range []struct {
		label string
		fn    func(a, b []byte) int
	}{{"naive_eq", naiveWrap}, {"ConstantTimeCompare", subtle.ConstantTimeCompare}} {
		out := fmt.Sprintf("   %-20s", v.label)
		for _, k := range []int{0, 8, 16, 31} {
			out += fmt.Sprintf(" k=%d:%.2f", k, avgNs(v.fn, candidate(k), 3000000))
		}
		fmt.Println(out)
	}
	_ = sink
	fmt.Println("   (naive trends up with k; ConstantTimeCompare flat. subtle also has")
	fmt.Println("    ConstantTimeSelect/ByteEq -- constant-time is a category, not one func.)")
	fmt.Println()
}

func partC() {
	fmt.Println("C. Rate limiting: attempts-to-first-success and effective limit")
	const listN, correctAt, configured = 1000, 500, 10

	run := func(mode string, workers, sourceIPs int) (int, bool) {
		if workers == 0 {
			workers = 1
		}
		if sourceIPs == 0 {
			sourceIPs = 1
		}
		allowed := 0
		reached := false
		buckets := map[interface{}]int{}
		for i := 1; i <= listN; i++ {
			ip := i % sourceIPs
			permitted := false
			switch mode {
			case "off":
				permitted = true
			case "redis_token_bucket":
				if _, ok := buckets["account"]; !ok {
					buckets["account"] = configured
				}
				if buckets["account"] > 0 {
					buckets["account"]--
					permitted = true
				}
			case "inproc":
				key := [2]int{0, i % workers}
				if _, ok := buckets[key]; !ok {
					buckets[key] = configured
				}
				if buckets[key] > 0 {
					buckets[key]--
					permitted = true
				}
			case "ip_keyed":
				if _, ok := buckets[ip]; !ok {
					buckets[ip] = configured
				}
				if buckets[ip] > 0 {
					buckets[ip]--
					permitted = true
				}
			}
			if permitted {
				allowed++
				if i == correctAt {
					reached = true
				}
			}
		}
		return allowed, reached
	}

	type row struct {
		mode              string
		workers, ips      int
		note              string
	}
	for _, r := range []row{
		{"off", 1, 1, "no limit"},
		{"redis_token_bucket", 1, 1, "shared bucket, configured=10"},
		{"inproc", 1, 1, "in-proc, 1 worker"},
		{"inproc", 4, 1, "in-proc, 4 workers -> effective 4x"},
		{"ip_keyed", 1, 50, "IP-keyed, attacker uses 50 IPs"},
	} {
		allowed, reached := run(r.mode, r.workers, r.ips)
		msg := "password NOT reached"
		if reached {
			msg = "reached password"
		}
		fmt.Printf("   %-18s %-34s allowed=%-4d %s\n", r.mode, r.note, allowed, msg)
	}
	fmt.Printf("\n   effective/configured: inproc workers=4 allows ~%d vs configured %d -> 4x.\n",
		4*configured, configured)
	fmt.Println("   IP-keyed with 50 IPs lets the password through -> keying on IP is a fake fix.")
	fmt.Println()
}

func main() {
	fmt.Println("Layer 7 · Topic 8 — hash cost, timing signal, rate limiting\n")
	partA()
	partB()
	partC()
	fmt.Println("Takeaway: password hash must be SLOW (argon2id), a secret compare " +
		"CONSTANT-TIME, and a rate limit keyed on the account with SHARED state.")
}

File: 07-security/08-crypto-and-rate-limiting/nodejs/crypto_ratelimit.js
// Layer 7 · Topic 8 — Crypto hygiene and rate limiting (Node.js).
//
// One command, no deps: `node crypto_ratelimit.js`. Node's constant-time
// primitive is crypto.timingSafeEqual, and it THROWS if the two buffers differ
// in length -- which forces you to think about the length leak every
// constant-time compare has (comparing fixed-length HASHES removes it). argon2
// and bcrypt are native modules; the built-in memory-hard KDF is scrypt, used
// here for Part A. Three parts, measured at runtime.
//
// What to look for: scrypt is ~5-6 orders slower per verify than sha256; the
// naive compare's ns/op climbs with matching bytes while timingSafeEqual stays
// flat; in-proc(workers=4) enforces ~4x the configured limit; IP-keyed with 50
// IPs stops nothing.

const crypto = require('node:crypto');

function partA() {
  console.log('A. Hash cost (verifications/sec, measured)');
  const pw = Buffer.from('correct horse battery staple');

  let reps = 300000;
  let t0 = process.hrtime.bigint();
  for (let i = 0; i < reps; i++) crypto.createHash('sha256').update(pw).digest();
  const shaVps = reps / (Number(process.hrtime.bigint() - t0) / 1e9);
  console.log(`   sha256          ${shaVps.toFixed(0).padStart(14)} verify/sec`);

  // scrypt at a deliberately costly setting (N=2^15) -- the built-in slow KDF.
  const salt = crypto.randomBytes(16);
  reps = 50;
  t0 = process.hrtime.bigint();
  for (let i = 0; i < reps; i++) crypto.scryptSync(pw, salt, 32, { N: 32768, r: 8, p: 1, maxmem: 64 * 1024 * 1024 });
  const scVps = reps / (Number(process.hrtime.bigint() - t0) / 1e9);
  console.log(`   scrypt(N=2^15)  ${scVps.toFixed(1).padStart(14)} verify/sec`);

  const N = 10000, K = 1000000;
  console.log(`   crack-time model: attacker rig N=${N}x, list K=${K} candidates`);
  console.log(`      sha256: ${(K / (shaVps * N)).toFixed(6)} s to first crack`);
  console.log(`      scrypt: ${(K / (scVps * N)).toFixed(1)} s to first crack  -- ~${(shaVps / scVps).toFixed(0)}x slower per verify`);
  console.log('   (argon2id is the OWASP first choice; it needs a native module here.)\n');
}

let sink = 0;
function naiveEq(a, b) {
  if (a.length !== b.length) return 0;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return 0; // short-circuit
  return 1;
}

function partB() {
  console.log('B. Timing signal: naive short-circuit vs constant-time');
  const secret = crypto.randomBytes(32);
  const candidate = (matching) => {
    const c = Buffer.from(crypto.randomBytes(32));
    secret.copy(c, 0, 0, matching);
    if (matching < 32) c[matching] = secret[matching] ^ 0xff;
    return c;
  };
  const tse = (a, b) => (crypto.timingSafeEqual(a, b) ? 1 : 0);
  const avgNs = (fn, cand, reps) => {
    const t0 = process.hrtime.bigint();
    for (let i = 0; i < reps; i++) sink += fn(secret, cand);
    return Number(process.hrtime.bigint() - t0) / reps;
  };
  console.log('   matching leading bytes ->        avg ns/op');
  for (const [label, fn] of [['naive_eq', naiveEq], ['timingSafeEqual', tse]]) {
    let out = `   ${label.padEnd(20)}`;
    for (const k of [0, 8, 16, 31]) out += ` k=${k}:${avgNs(fn, candidate(k), 1000000).toFixed(2)}`;
    console.log(out);
  }
  console.log('   (naive trends up with k; timingSafeEqual flat. It THROWS on unequal');
  console.log('    lengths -- compare fixed-length hashes to avoid the length leak.)\n');
}

function partC() {
  console.log('C. Rate limiting: attempts-to-first-success and effective limit');
  const LIST = 1000, CORRECT_AT = 500, CONFIGURED = 10;
  const run = (mode, workers = 1, sourceIPs = 1) => {
    let allowed = 0, reached = false;
    const buckets = new Map();
    for (let i = 1; i <= LIST; i++) {
      const ip = i % sourceIPs;
      let key, permitted = false;
      if (mode === 'off') permitted = true;
      else {
        if (mode === 'redis_token_bucket') key = 'account';
        else if (mode === 'inproc') key = `w${i % workers}`;
        else key = `ip${ip}`;
        if (!buckets.has(key)) buckets.set(key, CONFIGURED);
        if (buckets.get(key) > 0) { buckets.set(key, buckets.get(key) - 1); permitted = true; }
      }
      if (permitted) { allowed++; if (i === CORRECT_AT) reached = true; }
    }
    return [allowed, reached];
  };
  const rows = [
    ['off', 1, 1, 'no limit'],
    ['redis_token_bucket', 1, 1, 'shared bucket, configured=10'],
    ['inproc', 1, 1, 'in-proc, 1 worker'],
    ['inproc', 4, 1, 'in-proc, 4 workers -> effective 4x'],
    ['ip_keyed', 1, 50, 'IP-keyed, attacker uses 50 IPs'],
  ];
  for (const [mode, w, ips, note] of rows) {
    const [allowed, reached] = run(mode, w, ips);
    console.log(`   ${mode.padEnd(18)} ${note.padEnd(34)} allowed=${String(allowed).padEnd(4)} ${reached ? 'reached password' : 'password NOT reached'}`);
  }
  console.log(`\n   effective/configured: inproc workers=4 allows ~${4 * CONFIGURED} vs configured ${CONFIGURED} -> 4x.`);
  console.log('   IP-keyed with 50 IPs lets the password through -> keying on IP is a fake fix.\n');
}

console.log('Layer 7 · Topic 8 — hash cost, timing signal, rate limiting\n');
partA(); partB(); partC();
if (sink === -1) console.log(sink); // keep sink live
console.log('Takeaway: password hash must be SLOW, a secret compare CONSTANT-TIME, and ' +
  'a rate limit keyed on the account with SHARED state.');
